[Op] Test and fixes for shape variables

MasterJH5574 opened the pull request:
This PR brings unit tests for cases like “the shape of a tensor is a shape variable”, which are not covered by previous PRs listed #62.

By adding such tests, a few bugs of previous structure info deduction got revealed. The reason of those bugs is mainly the under consideration of shape variables. And this PR fixes them accordingly.

MasterJH5574 committed:
[Op] Test and fixes for shape variables

tqchen commented on src/relax/op/op_common.cc:
```diff
@@ -0,0 +1,121 @@
+                                               const Array<Integer>& axes, const String op_name) {
+  std::vector<bool> appeared_dims_set;
+  std::vector<int> axes_non_neg;
+  appeared_dims_set.resize(ndim);
```
This value should be initialized to false?

tqchen replied:
`appeared_dims_set.resize(ndim, false);`

Consider adding a regression test case

MasterJH5574 replied:
Std containers will uses `false` by default. But it’s also good to give it in an explicit way.

What do you mean by regression test here?

tqchen replied:
if it is by default then no regression is necessary

tqchen commented on src/relax/op/op_common.cc:
```diff
@@ -0,0 +1,121 @@
+
+Array<TensorStructInfo> GetInputTensorStructInfo(const Call& call, const BlockBuilder& ctx,
+                                                 const Array<String>& input_names,
+                                                 const String& op_name) {
```
Consider directly pass in const Op& here, to avoid adding refcount of string when there is no error

MasterJH5574 replied:
I realize that we can get that op from `call->op`. So we don’t even need to pass it in.

MasterJH5574 replied:
> ideally fixes should be applied back to each PR so they are self-contained. Given there is inter-dep between the PR, let us focus on getting this PR right and merge.

@tqchen I still want to back the fixes to their own PR. Will take some time to do that well.

tqchen commented on src/relax/op/op_common.cc:
```diff
@@ -0,0 +1,121 @@
+namespace relax {
+
+Array<TensorStructInfo> GetInputTensorStructInfo(const Call& call, const BlockBuilder& ctx,
+                                                 const Array<String>& input_names,
```
Avoid eagerly construct input_names, only construct when there is a need to do error reporting

tqchen commented on src/relax/op/op_common.cc:
```diff
@@ -0,0 +1,121 @@
+Optional<Array<PrimExpr>> InferBinaryBroadcastShape(const Call& call, const BlockBuilder& ctx,
+                                                    const Array<PrimExpr>& lhs_shape,
+                                                    const Array<PrimExpr>& rhs_shape,
+                                                    const String& op_name) {
```
Pass in const Op&

tqchen commented on src/relax/op/op_common.cc:
```diff
@@ -0,0 +1,121 @@
+  return Array<PrimExpr>(output_shape.rbegin(), output_shape.rend());
+}
+
+std::vector<int> CheckAxesInRangeNonRepetitive(const Call& call, const BlockBuilder& ctx, int ndim,
```
Add ICHECK to require ndim to be known

tqchen replied:
Document the requirement at the interface

tqchen replied:
Add a regression test

MasterJH5574 replied:
Great point. Updated and will push soon.

One point, I’m not sure how to add the regression test in this case, given it is not a public API. Perhaps ICHECK is a good enough gatekeeper?

tqchen commented on src/relax/op/tensor/reduce.cc:
```diff
@@ -0,0 +1,117 @@
+
+  Array<PrimExpr> out_shape;
+  out_shape.reserve(out_ndim);
+  for (int i = 0; i < data_sinfo->ndim; ++i) {
```
Add comment: 

// axes=None, keepdims = True, all values are 1
// axes is not None, keepdims=True, values matches, return 1
// axes is not None, keepdims=False, skip axis

MasterJH5574 replied:
Added in #73.

tqchen commented on tests/python/relax/test_op_reduce.py:
```diff
@@ -0,0 +1,190 @@
+
+
+def test_op_correctness():
+    x = relax.Var("x", R.Tensor((2, 3, 4, 5), "float32"))
```
test case coverage, Tensor,

- dtype=None
- dtype=float64, int32

MasterJH5574 replied:
We have them already.

tqchen commented on tests/python/relax/test_op_nn_convolution.py:
```diff
@@ -189,6 +189,35 @@ def test_conv2d_infer_struct_info_symbolic():
 
+def test_conv2d_infer_struct_info_shape_var():
+    bb = relax.BlockBuilder()
+    s0 = relax.Var("s", relax.ShapeStructInfo(ndim=4))
```
Common mix precision testcase:

test case coverage, Tensor,
- input dtype=f16, outdtype=f32
- input dtype=i8, outdtype=i32
- intput dtype=None, have outdtype

tqchen commented on tests/python/relax/test_op_nn.py:
```diff
@@ -151,5 +193,982 @@ def test_softmax_infer_struct_info_wrong_input_type():
+
+def test_layer_norm_infer_struct_info_more_input_dtype():
+    bb = relax.BlockBuilder()
+    x0 = relax.Var("x", R.Tensor((2, 3, 4, 5), "int8"))
```
BatchNorm would require dtype to be floating pt, int won't work. Consider add check instead.

Cross check other NN operators that involves real arithmetics

SiriusNEO commented on src/relax/op/nn/nn.cc:
```diff
@@ -68,5 +60,338 @@ TVM_REGISTER_OP("relax.nn.softmax")
+  int rhs_ndim = rhs_sinfo->ndim;
+  if (lhs_ndim == 0 || rhs_ndim == 0) {
+    ctx->ReportFatal(Diagnostic::Error(call)
+                     << "Matmul requires both input to have at least 1 dimension. However, "
```
```suggestion
                     << "Matmul requires both inputs to have at least 1 dimension. However, "
```

tqchen commented:
High-level comments:

- For operators that have non-linear ops which involves real numbers (e.g. exp, sqrt), we need to restrict their type to floating pt. Add is_float() check to the input. Would be good to have regression test cases
- Consider common mix precision use-cases, i8-accumulate i32, fp16- accumulate fp32
- Operator error message: avoid eagerly construct strings for error reporting purpose, pass in Op and lazily construct the error message.
- When it involves ndim, add ICHECK to avoid invariance.

tqchen commented:
ideally fixes should be applied back to each PR so they are self-contained. Given there is inter-dep between the PR, let us focus on getting this PR right and merge.

SiriusNEO commented on src/relax/op/nn/convolution.cc:
```diff
@@ -91,7 +94,7 @@ StructInfo InferStructInfoConv2d(const Call& call, const BlockBuilder& ctx) {
   PrimExpr input_channel_data = data_NCHW_shape[1];
   PrimExpr input_channel_kernel = weight_OIHW_shape[1];
-  if (analyzer->CanProve(input_channel_data - input_channel_kernel != 0)) {
+  if (analyzer->CanProve(input_channel_data != input_channel_kernel)) {
```
Would it be better if we use `CanProveEqual` here?

MasterJH5574 replied:
Generally there are three different cases:
1. Cases that we can prove the equality, like `2 == 2`, `n == n`.
2. Cases that we cannot prove neither the equality or the inequality, like `m = n?`.
3. Cases that we can prove the inequality, like `3 != 4`, `n != n + 2`.

To do best effort inference, we want to report error only when we can prove the inequality. That is to say, as long as we cannot prove the inequality, we should assume they are good. So here (and also for comments below) we use `CanProve(a != b)` instead of `CanProveEqual(a, b)`.

SiriusNEO commented on src/relax/op/nn/nn.cc:
```diff
@@ -68,5 +60,338 @@ TVM_REGISTER_OP("relax.nn.softmax")
+  arith::Analyzer* analyzer = ctx->GetAnalyzer();
+  PrimExpr lhs_reduction_length = lhs_shape->values[lhs_sinfo->ndim - 1];
+  PrimExpr rhs_reduction_length = rhs_shape->values[rhs_ndim - 2];
+  if (analyzer->CanProve(lhs_reduction_length != rhs_reduction_length)) {
```
`!CanProveEqual`

SiriusNEO commented on src/relax/op/nn/nn.cc:
```diff
@@ -68,5 +60,338 @@ TVM_REGISTER_OP("relax.nn.softmax")
+  }
+
+  arith::Analyzer* analyzer = ctx->GetAnalyzer();
+  if (analyzer->CanProve(pred_shape->values[0] != tgt_shape->values[0]) ||
+      analyzer->CanProve(pred_shape->values[1] != tgt_shape->values[1])) {
```
ditto

MasterJH5574 commented:
> ideally fixes should be applied back to each PR so they are self-contained. Given there is inter-dep between the PR, let us focus on getting this PR right and merge.

@tqchen I still want to back the fixes to their own PR. Will take some time to do that well.

Ubospica commented on python/tvm/relax/op/nn/nn.py:
```diff
@@ -326,3 +326,258 @@ def softmax(data: Expr, axis: int = -1) -> Expr:
+        The beta offset factor.
+
+    moving_mean : relax.Expr
+        Running mean of input,
```
```suggestion
        Running mean of input.
```

Ubospica commented on src/relax/op/op_common.h:
```diff
@@ -40,6 +40,19 @@ namespace relax {
 
+/*!
+ * \brief Get the tensor struct info of the operator input.
+ * \param call call The Call of the binary tensor operators.
```
```suggestion
 * \param call The Call of the tensor operator.
```

Ubospica commented on src/relax/op/nn/nn.cc:
```diff
@@ -68,5 +60,338 @@ TVM_REGISTER_OP("relax.nn.softmax")
+    }
+  }
+
+  std::vector<Array<PrimExpr>> axis_lengths;
```
Would using `Array<Array<PrimExpr>>` be better here?

MasterJH5574 replied:
No.

For one reason, constructing an Array by `push_back` is more time-consuming than constructing a std::vector in the same way, since each `push_back` to an Array does more than just appending a value.

For the other reason, `axis_lengths` is only used internally and won’t be returned and exposed in public API or be passed in FFI.

MasterJH5574 commented:
> High-level comments:
> 
> * For operators that have non-linear ops which involves real numbers (e.g. exp, sqrt), we need to restrict their type to floating pt. Add is_float() check to the input. Would be good to have regression test cases
> * Consider common mix precision use-cases, i8-accumulate i32, fp16- accumulate fp32
> * Operator error message: avoid eagerly construct strings for error reporting purpose, pass in Op and lazily construct the error message.
> * When it involves ndim, add ICHECK to avoid invariance.

@tqchen Addressed P3 and P4 in #72. Just rebased this PR with the latest fixes. Will resolve P1 and P2 in this PR after lunch.

MasterJH5574 committed:
Address comments: mixed precision

MasterJH5574 committed:
Address comments: Require float dtype

MasterJH5574 commented:
> > High-level comments:
> > 
> > * For operators that have non-linear ops which involves real numbers (e.g. exp, sqrt), we need to restrict their type to floating pt. Add is_float() check to the input. Would be good to have regression test cases
> > * Consider common mix precision use-cases, i8-accumulate i32, fp16- accumulate fp32
> > * Operator error message: avoid eagerly construct strings for error reporting purpose, pass in Op and lazily construct the error message.
> > * When it involves ndim, add ICHECK to avoid invariance.
> 
> @tqchen Addressed P3 and P4 in #72. Just rebased this PR with the latest fixes. Will resolve P1 and P2 in this PR after lunch.

Updated P1 and P2 with regression tests.

MasterJH5574 commented:
Depends on #72 and #73

MasterJH5574 commented:
Now depends on #73
